modules/core/src/test/scala/org/tesselation/trust/DataGenerator.scala

```diff
@@ -0,0 +1,68 @@
+package org.tesselation.trust
+
+import cats.effect.{Async, Ref}
+import cats.syntax.all._
+
+import scala.util.Random
+
+import org.tesselation.infrastructure.trust.{TrustEdge, TrustNode}
+
+class DataGenerator[F[_]: Async] {
+  private final val randomEffect: Ref[F, Random] = Ref.unsafe(new Random())
+
+  def randomEdgeLogic(distance: Double): F[Boolean] =
+    for {
+      random <- randomEffect.get
+    } yield random.nextDouble() > distance && random.nextDouble() < 0.5
+
+  def randomEdge(logic: Double => F[Boolean] = randomEdgeLogic)(n: TrustNode, n2: TrustNode) =
+    for {
+      random <- randomEffect.get
+      result <- logic(n.id.toDouble)
+    } yield
+      if (result) {
+        val trustZeroToOne = random.nextDouble()
+        Some(TrustEdge(n.id, n2.id, 2 * (trustZeroToOne - 0.5), isLabel = true))
+      } else None
+
+  def seedCliqueLogic(maxSeedNodeIdx: Int = 1)(id: Double): F[Boolean] =
+    for {
+      _ <- randomEffect.get
+    } yield
+      if (id <= maxSeedNodeIdx) true
+      else false
+
+  def cliqueEdge(logic: Double => F[Boolean] = seedCliqueLogic())(n: TrustNode, n2: TrustNode): F[Option[TrustEdge]] =
+    for {
+      random <- randomEffect.get
+      result <- logic(n.distance(n2))
+    } yield
+      if (result) Some(TrustEdge(n.id, n2.id, 1.0, isLabel = true))
+      else {
+        val trustZeroToOne = random.nextDouble()
+        Some(TrustEdge(n.id, n2.id, 2 * (trustZeroToOne - 0.5)))
+      }
+
+  def generateData(
+    numNodes: Int = 30,
+    edgeLogic: (TrustNode, TrustNode) => F[Option[TrustEdge]] = randomEdge()
+  ): F[List[TrustNode]] =
+    for {
+      random <- randomEffect.get
+      nodes = (0 until numNodes).toList.map(id => TrustNode(id, random.nextDouble(), random.nextDouble()))
+      nodesWithEdges = nodes.traverse { n =>
+        val edges = nodes.filterNot(_.id == n.id).traverse { n2 =>
+          edgeLogic(n, n2)
+        }
+        edges.map(e => n.copy(edges = e.flatten))
+      }
+      res <- nodesWithEdges
+    } yield res
+
+  def bipartiteEdge(logic: Double => F[Boolean] = seedCliqueLogic())(n: TrustNode, n2: TrustNode) =
+    for {
+      result <- logic(n.id.toDouble)
+    } yield
+      if (result) Some(TrustEdge(n.id, n2.id, 1.0, isLabel = true))
+      else Some(TrustEdge(n.id, n2.id, -1.0))
```

Comment:
for-comprehension is not needed if you are not `flatMapping`

```scala
logic(n.id.toDouble).map { result =>
  if (result) Some(TrustEdge(n.id, n2.id, 1.0, isLabel = true),
  else Some(TrustEdge(n.id, n2.id, -1.0)
}
```

Reply:
Same, will update this but didn't modify any of this code.